Property flags and descriptors (#219)
javascript-tutorial/pt.javascript.info: 1-js/07-object-properties/01-property-descriptors/article.md

 */
```

> Compare isso com o `user.name` "criado normalmente" acima: agora todos os sinalizadores são falsos. Se não é isso que queremos, então é melhor setá-los como `true` no `descriptor`.  

Reviewer: ```suggestion
Compare isso com o `user.name` "criado normalmente" acima: agora todos os sinalizadores são falsos. Se não é isso que queremos, então é melhor configurá-los como `true` no `descriptor`.  
```

> */!*

Reviewer: ```suggestion
Error: Cannot assign to read only property 'name'... (Erro: não é possível a atribuição à variável de apenas leitura 'name'...)
```

```

Agora, ninguém pode alterar o nome do nosso usuário, a não ser que eles apliquem seus próprios `defineProperty` para sobrescrever o nosso.

```smart header="Erros aparecem apenas em strict mode"

> No modo não-estrito, os erros não ocorrem quando gravando em propriedades não-graváveis, etc. Mas a operação ainda não terá sucesso. Ações que violam os sinalizadores são apenas ignoradas silenciosamentes em modo não-estrito. 

Reviewer: ```suggestion
No modo não-estrito, os erros não ocorrem quando escrevendo em propriedades não-graváveis, etc. Mas a operação ainda assim não terá sucesso. Ações que violam os sinalizadores são apenas ignoradas silenciosamentes em modo não-estrito. 
```

```

Aqui está o mesmo exemplo, mas a propriedade é criada do zero.

```js run
let user = { };

Object.defineProperty(user, "name", {
*!*
  value: "John",

>   // para novas proprieades, precisamos explicitamente listar o que é true

Reviewer: ```suggestion
  // para novas proprieades, precisamos explicitamente de listar o que é true
```

*/!*
```

Agora, ninguém pode alterar o nome do nosso usuário, a não ser que eles apliquem seus próprios `defineProperty` para sobrescrever o nosso.

```smart header="Erros aparecem apenas em strict mode"
No modo não-estrito, os erros não ocorrem quando gravando em propriedades não-graváveis, etc. Mas a operação ainda não terá sucesso. Ações que violam os sinalizadores são apenas ignoradas silenciosamentes em modo não-estrito. 
```

Aqui está o mesmo exemplo, mas a propriedade é criada do zero.

```js run
let user = { };

Object.defineProperty(user, "name", {
*!*
  value: "John",
  // para novas proprieades, precisamos explicitamente listar o que é true
  enumerable: true,
  configurable: true
*/!*
});

alert(user.name); // John
user.name = "Alice"; // Erro
```

## Não-enumerável

Agora, vamos adicionar um `toString` customizado ao `user`.

> Normalmente, um `toString` embutido em objetos é não-enumerável, e não aparece em `for...in`. Mas se nós adicionarmos um `toString` por nós mesmos, então por padrão ele aparece em `for...in`, desta forma:

Reviewer: ```suggestion
Normalmente, um `toString` embutido em objetos é não-enumerável, e não aparece em `for..in`. Mas se nós adicionarmos um `toString` por nós mesmos, então por padrão ele aparece em `for..in`, desta forma:
```

  }
};

// Por padrão, ambas as nossas propriedades são listadas:
for (let key in user) alert(key); // name, toString
```

> Se nós não gostarmos disso, então podemos setar `enumerable:false`. Então ela não vai aparecer no loop `for...in`, assim como as propriedades embutidas:

Reviewer: ```suggestion
Se nós não gostarmos disso, então podemos configurar `enumerable:false`. Então ela não vai aparecer no loop `for..in`, tal como as propriedades embutidas:
```

});

*!*
// Agora nosso toString desaparece:
*/!*
for (let key in user) alert(key); // name
```

Propriedades não-enumeráveis também são excluídas de `Object.keys`:

```js
alert(Object.keys(user)); // name
```

## Não-configurável

> O sinalizador não-configurável (`configurable:false`) algumas vezes é predefinido por objetos e propriedades embutidas.

Reviewer: ```suggestion
O sinalizador não-configurável (`configurable:false`) algumas vezes está predefinido para objetos e propriedades embutidas.
```

> O sinalizador não-configurável (`configurable:false`) algumas vezes é predefinido por objetos e propriedades embutidas.

Reviewer: ```suggestion
O sinalizador não-configurável (`configurable:false`) algumas vezes está predefinido para objetos e propriedades embutidas.
```

}
*/
```
Então, um programador é impossibilitado de mudar o valor de `Math.PI` ou sobrescrevê-lo.

```js run
Math.PI = 3; // Erro

// deletar Math.PI também não irá funcionar
```

Deixar uma propriedade não-configurável, é um caminho só de ida. Nós não podemos alterar isso novamente com `defineProperty`.

> Para ser preciso, não-configurabilidade impões várias restrições em `defineProperty`:

Reviewer: ```suggestion
Para ser preciso, a não-configurabilidade impões várias restrições a `defineProperty`:
```

1. Não poder mudar o sinalizador `configurable`.

> 1. Não poder mudar o sinalizador `enumerable`.

Reviewer: ```suggestion
2. Não poder mudar o sinalizador `enumerable`.
```

3. Não poder mudar `writable: false` para `true` (o contrário funciona).
4. Não poder mudar `get/set` por um acessador de propriedade (mas pode atribuí-los se ausente).

> **A ideia de "configurable: false" é para prevenir mudanças de sinalizadores de propriedades e sua eliminação, enquanto permite alterar seu valor.**

Reviewer: ```suggestion
**A ideia de "configurable: false" é para prevenir a mudança de sinalizadores de propriedades e a sua eliminação, enquanto permite alterar o seu valor.**
```

> 

Reviewer: ```suggestion
// Erro, porque configurable: false
```

  configurable: false
});

> // não será possível alterar user.name ou seus sinalizadores

Reviewer: ```suggestion
// não será possível alterar user.name ou os seus sinalizadores
```

});
```

> Então, nós podemos setar várias propriedades de uma vez.

Reviewer: ```suggestion
Então, nós podemos configurar várias propriedades de uma vez.
```

## Object.getOwnPropertyDescriptors

Para obter todos os sinalizadores de propriedade de uma vez, nós podemos usar o método [Object.getOwnPropertyDescriptors(obj)](mdn:js/Object/getOwnPropertyDescriptors).

> Juntamente com `Object.defineProperties` isso pode ser usado como um jeito "consciente-de-sinalizadores" de clonar objetos:

Reviewer: ```suggestion
Juntamente com `Object.defineProperties` isso pode ser usado como um jeito "incluindo-sinalizadores" de clonar objetos:
```

```js
let clone = Object.defineProperties({}, Object.getOwnPropertyDescriptors(obj));
```

Normalmente quando nós clonamos um objeto, nós usamos uma atribuição para copiar propriedades, desta forma:

```js
for (let key in user) {
  clone[key] = user[key]
}
```

> ...Mas isso não copia os sinalizadores. Então se nós quisermos um clone "melhor" então é preferível `Object.defineProperties`.

Reviewer: ```suggestion
...Mas isso não copia os sinalizadores. Assim, se nós quisermos um clone "melhor" então é preferível `Object.defineProperties`.
```

Outra diferença é que `for..in` ignora propriedades simbólicas, mas `Object.getOwnPropertyDescriptors` returna *todas* as propriedades descritoras, incluindo as simbólicas.

## Selando um objeto globalmente

Descritores de propriedade atuam no mesmo nível de propriedades individuais.

Também existem métodos que limitam o acesso ao objeto *inteiro*:

[Object.preventExtensions(obj)](mdn:js/Object/preventExtensions)
: Proíbe a adição de novas propriedades ao objeto.

[Object.seal(obj)](mdn:js/Object/seal)

> : Proíbe a adição/remoção de propriedades. Seta `configurable: false` para todas as propriedades existentes.

Reviewer: ```suggestion
: Proíbe a adição/remoção de propriedades. Coloca `configurable: false` para todas as propriedades existentes.
```

[Object.freeze(obj)](mdn:js/Object/freeze)

> : Proíbe adicionar/remover/alterar propriedades. Seta `configurable: false, writable: false` para todas as propriedades existentes.

Reviewer: ```suggestion
: Proíbe adicionar/remover/alterar propriedades. Coloca `configurable: false, writable: false` para todas as propriedades existentes.
```
